File: python/feedo/modules/storage.py
import httpx
import io
import time
from typing import Dict, Any, Optional
from eth_account.messages import encode_defunct
from eth_account import Account
from ..router import NodeRouter

class StorageModule:
    def __init__(self, router: NodeRouter, private_key: Optional[str] = None):
        self.router = router
        self.private_key = private_key

    def _build_auth_headers(self, method: str, path: str) -> Dict[str, str]:
        if not self.private_key:
            return {}
        account = Account.from_key(self.private_key)
        did = f"did:feedo:{account.address}"
        timestamp = str(int(time.time() * 1000))
        payload_str = f"FeedoAction:{method}:{path}:{timestamp}"
        message = encode_defunct(text=payload_str)
        signed_message = Account.sign_message(message, private_key=self.private_key)
        return {
            'X-Feedo-DID': did,
            'X-Feedo-Timestamp': timestamp,
            'X-Feedo-Signature': signed_message.signature.hex()
        }

    async def _request(self, method: str, path: str, json: Optional[Dict] = None, data: Any = None, files: Any = None) -> Any:
        base_url = await self.router.get_storage_node()
        url = f"{base_url}{path}"
        headers = self._build_auth_headers(method, path)

        async with httpx.AsyncClient() as client:
            try:
                response = await client.request(method, url, json=json, data=data, files=files, headers=headers)
                response.raise_for_status()
                if "application/json" in response.headers.get("content-type", ""):
                    return response.json()
                return response.content
            except Exception:
                print(f"Storage request failed on {base_url}, finding new node...")
                self.router.invalidate_storage_node()
                base_url = await self.router.get_storage_node()
                url = f"{base_url}{path}"
                headers = self._build_auth_headers(method, path)
                response = await client.request(method, url, json=json, data=data, files=files, headers=headers)
                response.raise_for_status()
                if "application/json" in response.headers.get("content-type", ""):
                    return response.json()
                return response.content

    async def upload_file(self, file_path: str, filename: str = "file") -> str:
        """Upload a file from a local path. Returns the hash ID."""
        with open(file_path, "rb") as f:
            return await self.upload_bytes(f.read(), filename)

    async def _upload_single_chunk(self, data: bytes, filename: str) -> str:
        base_url = await self.router.get_storage_node()
        url = f"{base_url}/upload"
        headers = self._build_auth_headers("POST", "/upload")

        async with httpx.AsyncClient() as client:
            files = {"file": (filename, io.BytesIO(data), "application/octet-stream")}
            try:
                response = await client.post(url, files=files, headers=headers)
                response.raise_for_status()
                return response.text.strip()
            except Exception:
                print(f"Storage request failed on {base_url}, finding new node...")
                self.router.invalidate_storage_node()
                base_url = await self.router.get_storage_node()
                url = f"{base_url}/upload"
                headers = self._build_auth_headers("POST", "/upload")
                response = await client.post(url, files=files, headers=headers)
                response.raise_for_status()
                return response.text.strip()

    async def upload_bytes(self, data: bytes, filename: str = "file") -> str:
        """Upload raw bytes directly. Automatically chunks files > 5MB. Returns the hash ID."""
        CHUNK_SIZE = 5 * 1024 * 1024
        size = len(data)
        if size <= CHUNK_SIZE:
            return await self._upload_single_chunk(data, filename)

        import asyncio
        import json

        chunks = []
        offset = 0
        while offset < size:
            chunks.append(data[offset:offset + CHUNK_SIZE])
            offset += CHUNK_SIZE

        hashes = [None] * len(chunks)
        semaphore = asyncio.Semaphore(10)

        async def upload_worker(idx, chunk):
            async with semaphore:
                chunk_hash = await self._upload_single_chunk(chunk, f"{filename}.part{idx}")
                hashes[idx] = chunk_hash

        tasks = [upload_worker(i, c) for i, c in enumerate(chunks)]
        await asyncio.gather(*tasks)

        manifest = {
            "type": "feedo_manifest",
            "filename": filename,
            "total_size": size,
            "chunk_size": CHUNK_SIZE,
            "chunks": hashes
        }

        manifest_bytes = json.dumps(manifest).encode('utf-8')
        return await self._upload_single_chunk(manifest_bytes, "manifest.json")

    async def _download_single_chunk(self, hash_id: str) -> bytes:
        return await self._request("GET", f"/download/{hash_id}")

    async def download_file(self, hash_id: str) -> bytes:
        """Download a file by its hash ID. Automatically rebuilds chunked files. Returns raw bytes."""
        raw_data = await self._download_single_chunk(hash_id)

        if len(raw_data) < 1024 * 1024:
            import json
            import asyncio
            try:
                text = raw_data.decode('utf-8')
                manifest = json.loads(text)
                if manifest.get("type") == "feedo_manifest" and isinstance(manifest.get("chunks"), list):
                    semaphore = asyncio.Semaphore(10)
                    chunks_data = [None] * len(manifest["chunks"])

                    async def download_worker(idx, h):
                        async with semaphore:
                            chunks_data[idx] = await self._download_single_chunk(h)

                    tasks = [download_worker(i, h) for i, h in enumerate(manifest["chunks"])]
                    await asyncio.gather(*tasks)

                    return b''.join(chunks_data)
            except Exception:
                pass

        return raw_data

    async def ingest_json(self, payload: Dict):
        return await self._request("POST", "/api/v1/ingest/post", json=payload)

    async def get_recent_files(self):
        return await self._request("GET", "/api/files/recent")

File: typescript/src/modules/storage.ts
import axios from 'axios';
import { NodeRouter } from '../router';
import { ethers } from 'ethers';

export class StorageModule {
    constructor(private router: NodeRouter, private privateKey?: string) {}

    private async request(method: string, path: string, data?: any, isMultipart: boolean = false) {
        let baseUrl = await this.router.getStorageNode();
        let url = `${baseUrl}${path}`;
        
        const headers: any = {};

        if (this.privateKey) {
            const wallet = new ethers.Wallet(this.privateKey);
            const did = `did:feedo:${wallet.address}`;
            const timestamp = Date.now().toString();
            const payload = `FeedoAction:${method}:${path}:${timestamp}`;
            const signature = await wallet.signMessage(payload);
            
            headers['X-Feedo-DID'] = did;
            headers['X-Feedo-Timestamp'] = timestamp;
            headers['X-Feedo-Signature'] = signature;
        }

        try {
            const response = await axios({ method, url, data, headers });
            return response.data;
        } catch (error: any) {
            console.warn(`Storage request failed on ${baseUrl}, trying to find a new node...`);
            this.router.invalidateStorageNode();
            baseUrl = await this.router.getStorageNode();
            url = `${baseUrl}${path}`;
            const retryResponse = await axios({ method, url, data, headers });
            return retryResponse.data;
        }
    }

    private async uploadSingleChunk(fileBlobOrBuffer: any, filename: string): Promise<string> {
        let finalData = fileBlobOrBuffer;
        if (typeof Buffer !== 'undefined' && Buffer.isBuffer(fileBlobOrBuffer)) {
            finalData = new Blob([fileBlobOrBuffer as any]);
        } else if (fileBlobOrBuffer instanceof Uint8Array) {
            finalData = new Blob([fileBlobOrBuffer as any]);
        }

        const formData = new FormData();
        formData.append('file', finalData, filename);
        
        let baseUrl = await this.router.getStorageNode();
        let url = `${baseUrl}/upload`;
        
        const headers: any = {};
        if (this.privateKey) {
            const wallet = new ethers.Wallet(this.privateKey);
            const did = `did:feedo:${wallet.address}`;
            const timestamp = Date.now().toString();
            const payload = `FeedoAction:POST:/upload:${timestamp}`;
            const signature = await wallet.signMessage(payload);
            
            headers['X-Feedo-DID'] = did;
            headers['X-Feedo-Timestamp'] = timestamp;
            headers['X-Feedo-Signature'] = signature;
        }

        try {
            const response = await fetch(url, { method: 'POST', headers, body: formData as any });
            if (!response.ok) throw new Error(await response.text());
            return await response.text();
        } catch (error: any) {
            console.warn(`Storage request failed on ${baseUrl}, trying to find a new node...`);
            this.router.invalidateStorageNode();
            baseUrl = await this.router.getStorageNode();
            url = `${baseUrl}/upload`;
            const retryResponse = await fetch(url, { method: 'POST', headers, body: formData as any });
            if (!retryResponse.ok) throw new Error(await retryResponse.text());
            return await retryResponse.text();
        }
    }

    async uploadFile(fileBlobOrBuffer: any, filename: string = 'file') {
        let size = 0;
        if (fileBlobOrBuffer.size !== undefined) size = fileBlobOrBuffer.size;
        else if (fileBlobOrBuffer.byteLength !== undefined) size = fileBlobOrBuffer.byteLength;
        else if (fileBlobOrBuffer.length !== undefined) size = fileBlobOrBuffer.length;

        const CHUNK_SIZE = 5 * 1024 * 1024; // 5 MB
        
        if (size <= CHUNK_SIZE) {
            return this.uploadSingleChunk(fileBlobOrBuffer, filename);
        }

        // Chunking logic
        const chunks: any[] = [];
        let offset = 0;
        while (offset < size) {
            let chunk;
            if (fileBlobOrBuffer.slice) {
                // Blob or Buffer
                chunk = fileBlobOrBuffer.slice(offset, offset + CHUNK_SIZE);
            } else if (fileBlobOrBuffer.subarray) {
                // Uint8Array
                chunk = fileBlobOrBuffer.subarray(offset, offset + CHUNK_SIZE);
            } else {
                throw new Error("Unsupported file type for chunking");
            }
            chunks.push(chunk);
            offset += CHUNK_SIZE;
        }

        // Upload chunks with promise pool
        const limit = 10;
        const hashes: string[] = new Array(chunks.length);
        let i = 0;
        const workers = new Array(limit).fill(0).map(async () => {
            while (i < chunks.length) {
                const index = i++;
                const chunkFilename = `${filename}.part${index}`;
                hashes[index] = await this.uploadSingleChunk(chunks[index], chunkFilename);
            }
        });
        await Promise.all(workers);

        // Create manifest
        const manifest = {
            type: "feedo_manifest",
            filename: filename,
            total_size: size,
            chunk_size: CHUNK_SIZE,
            chunks: hashes
        };

        const manifestString = JSON.stringify(manifest);
        let manifestData;
        if (typeof Blob !== 'undefined') {
            manifestData = new Blob([manifestString], { type: 'application/json' });
        } else {
            manifestData = Buffer.from(manifestString, 'utf-8');
        }

        return await this.uploadSingleChunk(manifestData, 'manifest.json');
    }

    private async downloadSingleChunk(hash: string): Promise<ArrayBuffer> {
        let baseUrl = await this.router.getStorageNode();
        let path = `/download/${encodeURIComponent(hash)}`;
        let url = `${baseUrl}${path}`;
        
        const headers: any = {};
        if (this.privateKey) {
            const wallet = new ethers.Wallet(this.privateKey);
            const did = `did:feedo:${wallet.address}`;
            const timestamp = Date.now().toString();
            const payload = `FeedoAction:GET:${path}:${timestamp}`;
            const signature = await wallet.signMessage(payload);
            
            headers['X-Feedo-DID'] = did;
            headers['X-Feedo-Timestamp'] = timestamp;
            headers['X-Feedo-Signature'] = signature;
        }
        
        try {
            const response = await axios.get(url, { responseType: 'arraybuffer', headers });
            return response.data;
        } catch (error: any) {
            console.warn(`Download failed on ${baseUrl}, trying new node...`);
            this.router.invalidateStorageNode();
            baseUrl = await this.router.getStorageNode();
            path = `/download/${encodeURIComponent(hash)}`;
            url = `${baseUrl}${path}`;
            const retryResponse = await axios.get(url, { responseType: 'arraybuffer', headers });
            return retryResponse.data;
        }
    }

    async downloadFile(hash: string) {
        const rawData = await this.downloadSingleChunk(hash);
        
        // If it's small, it might be a manifest
        if (rawData.byteLength < 1024 * 1024) { 
            try {
                const text = new TextDecoder().decode(rawData);
                const json = JSON.parse(text);
                if (json.type === 'feedo_manifest' && Array.isArray(json.chunks)) {
                    // Download chunks with promise pool
                    const limit = 10;
                    const chunks: ArrayBuffer[] = new Array(json.chunks.length);
                    let i = 0;
                    const workers = new Array(limit).fill(0).map(async () => {
                        while (i < json.chunks.length) {
                            const index = i++;
                            chunks[index] = await this.downloadSingleChunk(json.chunks[index]);
                        }
                    });
                    await Promise.all(workers);
                    
                    // Concatenate chunks
                    let totalLen = chunks.reduce((acc, c) => acc + c.byteLength, 0);
                    let result = new Uint8Array(totalLen);
                    let offset = 0;
                    for (let c of chunks) {
                        result.set(new Uint8Array(c), offset);
                        offset += c.byteLength;
                    }
                    return result.buffer;
                }
            } catch (e) {
                // Not a manifest or not JSON, fallback to returning raw data
            }
        }
        
        return rawData;
    }

    async ingestJson(payload: any) {
        return this.request('POST', '/api/v1/ingest/post', payload);
    }

    async getRecentFiles() {
        return this.request('GET', '/api/files/recent');
    }
}
